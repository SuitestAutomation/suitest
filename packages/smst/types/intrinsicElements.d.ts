type SingleNode = import('./unistTestLine').SingleNode;
type Node = import('./unistTestLine').Node;
type CodeBlockLanguage = import('./unistTestLine').CodeBlockLanguage;
type SingleEntryStatus = import('./unistTestLine').SingleEntryStatus;
type TestLineResultStatus = import('./unistTestLine').TestLineResultStatus;
type ElementOrTextChildren = string | SingleNode | undefined | ElementOrTextChildren[];
type ElementChildren = SingleNode | undefined | ElementChildren[];
type StringOrStrings = string | string[];

declare namespace JSX {
	interface ElementChildrenAttribute {
		children: {},  // specify children name to use
	}

	// This is return value of ANY jsx element
	// TypeScript can't infer correct type, unfortunately,
	// so will simply put it to any node at all
	// https://www.typescriptlang.org/docs/handbook/jsx.html#the-jsx-result-type
	type Element = Node;

	interface IntrinsicElements {
		// Plain text, should accept only string or other textual elements to make
		// it easier to format text
		text: {children: StringOrStrings},
		subject: {children: StringOrStrings},
		input: {children: StringOrStrings},
		code: {children: StringOrStrings},
		fragment: {children: ElementOrTextChildren},
		'code-block': {
			children: StringOrStrings,
			language?: CodeBlockLanguage,
		},
		prop: {
			name: ElementChildren,
			comparator?: string,
			expectedValue: ElementChildren,
			actualValue?: string | number,
			status?: SingleEntryStatus,
		},
		props: {children: ElementChildren},
		condition: {
			title: ElementChildren,
			children?: ElementChildren,
			status?: SingleEntryStatus,
		},
		'test-line': {
			title: ElementChildren,
			children?: ElementChildren,
			status?: TestLineResultStatus,
		},
		'test-line-result': {
			status: TestLineResultStatus,
			children: ElementChildren,
			message?: ElementChildren,
			screenshot?: string,
		},
	}
}
