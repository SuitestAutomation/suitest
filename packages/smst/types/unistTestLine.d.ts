export type TextNode = {
	type: 'text',
	value: string,
};

export type SubjectNode = {
	type: 'subject',
	value: string,
};

export type InputNode = {
	type: 'input',
	value: string,
};

export type CodeNode = {
	type: 'code',
	value: string,
};

export type InlineTextNode = TextNode | SubjectNode | InputNode | CodeNode;

export type CodeBlockLanguage = 'javascript' | 'brightscript';

export type CodeBlockNode = {
	type: 'code-block',
	language?: CodeBlockLanguage,
	value: string,
};

export type SingleEntryStatus = 'success' | 'fail';

export type InlinePropertyNode = {
	type: 'prop',
	contentType: 'inline',
	name: InlineTextNode[],
	comparator?: string,
	expectedValue: InlineTextNode[],
	actualValue?: string | number | boolean,
	status?: SingleEntryStatus,
};

export type CodePropertyNode = {
	type: 'prop',
	contentType: 'block',
	name: InlineTextNode[],
	comparator?: string,
	expectedValue: CodeBlockNode,
	status?: SingleEntryStatus,
};

export type PropertyNode = InlinePropertyNode | CodePropertyNode;

export type PropertiesNode = {
	type: 'props',
	children: PropertyNode[],
};

export type ConditionNode = {
	type: 'condition',
	title: InlineTextNode[],
	children: Array<PropertiesNode | ConditionNode>,
	status?: SingleEntryStatus,
};

export type TestLineResultStatus = 'success' | 'fatal' | 'fail' | 'warning' | 'exit' | 'excluded' | 'aborted';

export type TestLineNode = {
	type: 'test-line',
	title: InlineTextNode[],
	children: Array<PropertiesNode | ConditionNode>,
	status?: TestLineResultStatus,
};

export type TestLineResultNode = {
	type: 'test-line-result',
	status: TestLineResultStatus,
	children: TestLineNode[],
	message?: Array<InlineTextNode | LinkNode>,
	screenshot?: string,
	docs?: string,
};

export type LinkNode = {
	type: 'link',
	href: string,
	value: string,
};

export type SingleNode =
	| InlineTextNode
	| CodeBlockNode
	| PropertyNode
	| PropertiesNode
	| ConditionNode
	| TestLineNode
	| TestLineResultNode
	| LinkNode;

export type Node =  SingleNode | SingleNode[];

export type Verbosity = 'quiet' | 'normal' | 'verbose';
