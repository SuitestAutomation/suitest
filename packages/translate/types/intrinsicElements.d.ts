/// <reference path="./unistTestLine.d.ts" />

type ElementOrTextChildren = string | SingleNode | undefined | ElementOrTextChildren[];
type ElementChildren = SingleNode | undefined | ElementChildren[];
type StringOrStrings = string | string[];

declare namespace JSX {
	interface ElementChildrenAttribute {
		children: {},  // specify children name to use
	}

	// This is return value of ANY jsx element
	// TypeScript can't infer correct type, unfortunately,
	// so will simply put it to any node at all
	// https://www.typescriptlang.org/docs/handbook/jsx.html#the-jsx-result-type
	type Element = Node;

	interface IntrinsicElements {
		// Plain text, should accept only string or other textual elements to make
		// it easier to format text
		text: {children: StringOrStrings},
		bold: {children: StringOrStrings},
		emphasis: {children: StringOrStrings},
		code: {children: StringOrStrings},
		alert: {
			level: 'error' | 'fatal' | 'warning' | 'exit' | 'excluded',
			children: ElementOrTextChildren,
		},
		fragment: {children: ElementOrTextChildren},
		paragraph: {children: ElementOrTextChildren},
		cell: {children: ElementOrTextChildren},
		row: {children: ElementChildren},
		table: {children: ElementChildren, label?: ElementOrTextChildren},
		dictionary: {children: ElementChildren, label?: ElementOrTextChildren},
		'code-block': {children: StringOrStrings, label?: ElementOrTextChildren},
		condition: {
			title: ElementOrTextChildren,
			children?: ElementChildren,
		},
		'test-line': {
			title: ElementOrTextChildren,
			children?: ElementChildren,
		},
	}
}
