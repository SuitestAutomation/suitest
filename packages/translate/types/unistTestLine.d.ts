declare type PlainTextNode = {
	type: 'text',
	value: string,
};
declare type BoldTextNode = {
	type: 'bold',
	value: string,
};
declare type EmphasisTextNode = {
	type: 'emphasis',
	value: string,
};
declare type CodeTextNode = {
	type: 'code',
	value: string,
};
declare type TextNode = PlainTextNode | BoldTextNode | EmphasisTextNode | CodeTextNode;
declare type ParagraphNode = {
	type: 'paragraph',
	children: TextNode[],
};
declare type CellNode = {
	type: 'cell',
	children: TextNode[],
};
declare type RowNode = {
	type: 'row',
	children: CellNode[],
};
declare type TableNode = {
	type: 'table',
	label?: TextNode[],
	children: RowNode[],
};
declare type DictionaryRowNode = {
	type: 'row',
	children: [CellNode, CellNode],
};
declare type DictionaryNode = {
	type: 'dictionary',
	label?: TextNode[],
	children: DictionaryRowNode[],
};
declare type CodeBlockNode = {
	type: 'code-block',
	label?: TextNode[],
	value: string,
};
declare type AlertNode = {
	type: 'alert',
	level: 'error' | 'fatal' | 'fail' | 'warning' | 'exit' | 'excluded',
	children: TextNode[],
};
declare type SectionNode = ParagraphNode
	| TableNode
	| DictionaryNode
	| CodeBlockNode
	| ConditionNode
	| AlertNode;
declare type TestLineNode = {
	type: 'test-line',
	title: TextNode[],
	children: SectionNode[],
};
declare type ConditionNode = {
	type: 'condition',
	title: TextNode[],
	children: SectionNode[],
};
declare type SingleNode = ConditionNode | TestLineNode | SectionNode | CellNode | RowNode | TextNode;
declare type Node =  SingleNode | SingleNode[];
