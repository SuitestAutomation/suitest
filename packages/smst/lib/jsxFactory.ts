// Reference is needed so that whenever this file is imported by consumer, intrinsic types are always included
/// <reference path="../types/intrinsicElements.d.ts" />
import ub from 'unist-builder';
import {
	CodeBlockNode,
	ConditionNode,
	InlineTextNode,
	Node,
	PropertyNode,
	SingleNode,
	TestLineNode,
	TextNode,
	InlinePropertyNode,
	CodePropertyNode,
} from '../types/unistTestLine';

const plainTypes = ['text', 'code', 'subject', 'input'];
const isTextNode = (input?: JSX.SmstFlatNode): input is TextNode =>
	typeof input === 'object' && input !== null && plainTypes.includes(input.type);

type DeepArrayOrOne<T> = T | Array<T | DeepArrayOrOne<T>>;
export const flatten = <T extends any>(input: DeepArrayOrOne<T>): T[] => {
	if (Array.isArray(input)) {
		return input.reduce((acc: T[], item: DeepArrayOrOne<T>) => acc.concat(flatten(item)), []);
	}

	return [input];
};

/**
 * Normalize textual children - make them flat and merge sequential text items of same type
 */
function normalizePlainChildren
	<R extends SingleNode, K extends keyof JSX.IntrinsicElements = R['type']>
	(children: JSX.SmstFlatNode[], parentType: K): R[];
function normalizePlainChildren(
	children: JSX.SmstFlatNode[],
	parentType: keyof JSX.IntrinsicElements
): SingleNode[] {
	return children.reduce((output: SingleNode[], child) => {
		if (child === undefined || child === null) {
			return output;
		}

		if (typeof child === 'string' || typeof child === 'boolean' || typeof child === 'number') {
			// Wrap plain strings into text nodes
			child = ub(plainTypes.includes(parentType) ? parentType : 'text', String(child)) as InlineTextNode;
		}

		// Nested child is a textual node
		const lastChild = output[output.length - 1];

		if (isTextNode(child) && lastChild?.type === child.type) {
			// Textual node with same time as the previous one - merge them
			output.splice(output.length - 1, 1, {
				type: child.type,
				value: lastChild.value + child.value,
			});
		} else {
			output.push(child);
		}

		return output;
	}, []);
}

const processTextNode = (label: JSX.SmstNode): TextNode[] =>
	normalizePlainChildren<TextNode>(flatten([label]), 'text');

const processPropertyNode = (props: JSX.ElementsProps['prop']): InlinePropertyNode | CodePropertyNode => {
	const name = processTextNode(props.name);

	if (isCodeBlockNode(props.expectedValue)) {
		return ub('prop', {
			name,
			contentType: 'block',
			expectedValue: props.expectedValue,
			comparator: props.comparator,
			status: props.status,
		});
	}

	return ub('prop', {
		name,
		contentType: 'inline',
		expectedValue: processTextNode(props.expectedValue),
		actualValue: props.actualValue,
		comparator: props.comparator,
		status: props.status,
	});
};

const isPropChildren = (children: SingleNode[]): children is PropertyNode[] =>
	children.every(child => child.type === 'prop');

const isCodeBlockNode = (item: JSX.SmstNode): item is CodeBlockNode =>
	!Array.isArray(item) && typeof item === 'object' && item !== null && item.type === 'code-block';

const isTestLineNodeArray = (children: SingleNode[]): children is TestLineNode[] =>
	children.every(child => child.type === 'test-line');

/* istanbul ignore next */
const assertUnknownIntrinsicElementType = (type: never): never => {
	throw new Error(`Unknown intrinsic element type ${type}`);
};

/* istanbul ignore next */
export const assertUnknownSectionNode = (node: never): never => {
	throw new Error('Unknown node type: ' + JSON.stringify(node));
};

/* istanbul ignore next */
export const assertUnknownTextNode = (node: never): never => {
	throw new Error('Unknown plain text node: ' + JSON.stringify(node));
};

/**
 * A factory for JSX elements. Maps React-style JSX to unist-style nodes.
 * "any" is used for children and props because TypeScript can't infer JSX element types
 * anyway and always returns whatever is provided in JSX.Element type/interface.
 * Some validation is added to ensure correct data at least runtime.
 */
export function jsx
	<K extends keyof JSX.ElementsProps>
	(type: K, props: JSX.ElementsProps[K], ...children: Array<JSX.ElementsChildren[K]>): Node;
export function jsx(
	type: keyof JSX.ElementsProps,
	props: JSX.ElementsProps[keyof JSX.ElementsProps],
	...children: JSX.SmstNode[]
): Node {
	// Do not include empty children, to allow for ternary operations in JSX
	// Flatten the children - to support nested arrays in JSX and manage fragments
	const processedChildren = normalizePlainChildren(flatten(children), type);

	switch (type) {
		case 'text':
		case 'subject':
		case 'input':
		case 'code':
			return processedChildren[0];
		case 'fragment':
			return processedChildren;
		case 'code-block':
			const codeBlockProps = props as JSX.ElementsProps['code-block'];
			const codeBlockChildren = processedChildren[0] as CodeBlockNode;
			const language = codeBlockProps?.language ?? 'javascript';

			// Type casting because IntrinsicElements definition would not allow anything other then string
			return ub('code-block', {language}, codeBlockChildren.value) as CodeBlockNode;
		case 'prop':
			// props can't be null because it's defined in intrinsicElements
			return processPropertyNode(props as JSX.ElementsProps['prop']);
		case 'props':
			if (isPropChildren(processedChildren)) {
				return ub('props', processedChildren);
			}

			throw new TypeError('Props can only accept Prop as its child');
		case 'condition':
		case 'test-line':
			const testLineProps = props as JSX.ElementsProps['test-line'];
			// Type casting for props because of IntrinsicElements definition
			const params: Omit<TestLineNode, 'type' | 'children'> = {
				title: processTextNode(testLineProps.title),
			};

			if (testLineProps.status) {
				params.status = testLineProps.status;
			}

			return ub(type, params, processedChildren) as ConditionNode | TestLineNode;
		case 'test-line-result':
			const testLineResultProps = props as JSX.ElementsProps['test-line-result'];
			if (isTestLineNodeArray(processedChildren) && processedChildren.length === 1) {
				return ub('test-line-result', {
					status: testLineResultProps.status,
					message: processTextNode(testLineResultProps.message),
					screenshot: props?.screenshot,
				}, processedChildren);
			}

			throw new TypeError('TestLineResult expects a single child of type TestLine');
		default:
			/* istanbul ignore next */
			return assertUnknownIntrinsicElementType(type);
	}
}
