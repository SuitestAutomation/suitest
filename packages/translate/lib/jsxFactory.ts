import '../types/intrinsicElements.d.ts';
import * as assert from 'assert';
import * as ub from 'unist-builder';
import {
	ConditionNode,
	DictionaryNode,
	Node,
	TableNode,
	TestLineNode,
	TextFragmentNode,
	TextNode,
} from '../types/unistTestLine';

type Label = string | TextNode | TextFragmentNode;

const plainTypes = ['text-fragment', 'text', 'code', 'bold', 'emphasis'];

const buildTextFragment = (children: any[]): TextFragmentNode =>
	ub('text-fragment', children.map(child => {
		if (typeof  child === 'string') {
			return ub('text', child);
		}

		assert(plainTypes.includes(child.type), 'Text fragment can only accept textual children');

		return child;
	}).reduce((newChildren, child) => {
		// Recursively process nested text fragments
		const childrenToProcess = child.type === 'text-fragment'
			? (jsx('text-fragment', {}, ...child.children) as TextFragmentNode).children
			: [child];

		childrenToProcess.forEach((nestedChild: TextNode | string) => {
			if (typeof nestedChild === 'string') {
				nestedChild = ub('text', nestedChild);
			}

			const lastChild = newChildren[newChildren.length - 1];

			if (lastChild?.type === nestedChild.type) {
				lastChild.value += nestedChild.value;
			} else {
				newChildren.push(nestedChild);
			}
		});

		return newChildren;
	}, []));

const buildTable = (children: any[], label?: Label): TableNode => {
	children.forEach(child =>
		assert(child.type === 'row', 'Table can accepts only Row as it\'s child')
	);

	const tableLabel = typeof label === 'string' ? ub('text', label) : label;

	return ub('table', {label: tableLabel}, children);
};

const buildDictionary = (children: any[], label?: Label): DictionaryNode => {
	children.forEach(child => {
		assert(child.type === 'row', 'Dictionary can accepts only Row as it\'s child');
		assert(child.children.length === 2, 'Dictionary expects exactly 2 child nodes');
	});

	const dictLabel = typeof label === 'string' ? ub('text', label) : label;

	return ub('dictionary', {label: dictLabel}, children);
};

const assertUnknownIntrinsicElementType = (type: never): never => {
	throw new Error(`Unknown intrinsic element type ${type}`);
};

/**
 * A factory for JSX elements. Maps React-style JSX to unist-style nodes.
 * "any" is used for children and props because TypeScript can't infer JSX element types
 * anyway and always returns whatever is provided in JSX.Element type/interface.
 * Some validation is added to ensure correct data at least runtime.
 */
export const jsx = (type: Node['type'], props: {[key: string]: any}, ...children: any[]): Node => {
	// Do not include empty children, to allow for ternary operations in
	children = children.filter(Boolean);

	switch (type) {
		case 'text':
		case 'bold':
		case 'emphasis':
		case 'code':
			// Can have multiple string children, e.g. if got <text>some {variable}</text>
			// Validation of children is done statically by typescript
			return ub(type, children.join(''));
		case 'text-fragment':
			return buildTextFragment(children);
		case 'row':
			children.forEach(child =>
				assert(child.type === 'cell', 'Row can accepts only Cell as it\'s child')
			);

			return ub('row', children);
		case 'table':
			return buildTable(children, props?.label);
		case 'dictionary':
			return buildDictionary(children, props?.label);
		case 'cell':
		case 'paragraph':
		case 'warning-block':
		case 'error-block':
			// Pass children through test fragment logic to it's wrapped correctly
			return ub(type, buildTextFragment(children).children);
		case 'condition':
		case 'test-line':
			const title = typeof props.title === 'string' ? ub('text', props.title) : props.title;

			return ub(type, {title}, children) as ConditionNode | TestLineNode;
		case 'code-block':
			const codeBlockLabel = typeof props?.label === 'string' ? ub('text', props.label) : props?.label;

			// Code block is a special case - block that can receive only strings as children
			return ub('code-block', {label: codeBlockLabel}, children.join(''));
		default:
			return assertUnknownIntrinsicElementType(type);
	}
};
