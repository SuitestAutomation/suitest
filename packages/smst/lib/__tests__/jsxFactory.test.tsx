import {CodeBlockNode, CodeNode, InputNode, Node, SubjectNode, TextNode} from '../../types/unistTestLine';
import {jsx, flatten} from '../jsxFactory';

describe('jsxFactory', () => {
	it('should accept JSX node and generate correct unist elements out of it', () => {
		const plain: TextNode = {
			type: 'text',
			value: 'text',
		};
		expect(<text>text</text>).toEqual(plain);

		const subj: SubjectNode = {
			type: 'subject',
			value: 'text',
		};
		expect(<subject>text</subject>).toEqual(subj);

		const input: InputNode = {
			type: 'input',
			value: 'text',
		};
		expect(<input>text</input>).toEqual(input);

		const code: CodeNode = {
			type: 'code',
			value: 'text',
		};
		expect(<code>text</code>).toEqual(code);

		const codeBlock: CodeBlockNode = {
			type: 'code-block',
			language: 'javascript',
			value: 'test',
		};
		expect(<code-block>test</code-block>).toEqual(codeBlock);
		expect(<code-block language="javascript">test</code-block>).toEqual(codeBlock);

		// Has to produce empty text node instead of just skipping the element altogether
		const emptyText = {
			type: 'text',
			value: '',
		};
		expect(<text>{''}</text>).toEqual(emptyText);
	});

	it('should process complex text nodes', () => {
		const jsxResult = <fragment>some <subject>subject</subject> and <input>input</input> text</fragment>;
		const unist: Node[] = [
			{
				type: 'text',
				value: 'some ',
			},
			{
				type: 'subject',
				value: 'subject',
			},
			{
				type: 'text',
				value: ' and ',
			},
			{
				type: 'input',
				value: 'input',
			},
			{
				type: 'text',
				value: ' text',
			},
		];

		expect(jsxResult).toEqual(unist);
	});

	it('should filter out empty elements and merge same element types that are listed in sequence', () => {
		const jsxResult = <fragment>some {undefined} string <input>input</input></fragment>;
		const unist: Node[] = [
			{
				type: 'text',
				value: 'some  string ',
			},
			{
				type: 'input',
				value: 'input',
			},
		];

		expect(jsxResult).toEqual(unist);
	});

	it('should unfold any nested text fragment into a single one', () => {
		const jsxResult = <fragment>
			<fragment>Some test</fragment> <subject>another</subject>
		</fragment>;
		const unist: Node[] = [
			{
				type: 'text',
				value: 'Some test ',
			},
			{
				type: 'subject',
				value: 'another',
			},
		];

		expect(jsxResult).toEqual(unist);
	});

	it('should create correct Prop element', () => {
		expect(<prop
			name={<text>prop name</text>}
			expectedValue={<text>expected value</text>}
			actualValue="actual value"
			comparator="="
			status="success"
		/>).toMatchSnapshot();

		expect(<prop
			name={<text>prop name</text>}
			expectedValue={<code-block>expected value</code-block>}
			status="fail"
		/>).toMatchSnapshot();
	});

	it('should create correct Props element', () => {
		expect(<props><prop name={<text>test</text>} expectedValue={<text>test</text>}/></props>)
			.toMatchSnapshot();
	});

	it('should throw if props element receives anything other than row', () => {
		expect(() => <props><text>...</text></props>).toThrow();
	});

	it('should save title property for test line and condition', () => {
		expect(<test-line
			title={<text>Some title</text>}
			status="success"
		>
			<text>text</text>
		</test-line>).toEqual({
			type: 'test-line',
			title: [{type: 'text', value: 'Some title'}],
			status: 'success',
			children: [{type: 'text', value: 'text'}],
		});

		expect(<test-line title={<fragment>Some <subject>subject</subject> title</fragment>}>
			<text>text</text>
		</test-line>).toEqual({
			type: 'test-line',
			title: [
				{type: 'text', value: 'Some '},
				{type: 'subject', value: 'subject'},
				{type: 'text', value: ' title'},
			],
			children: [{type: 'text', value: 'text'}],
		});

		expect(<condition
			title={<text>Some title</text>}
			status="success"
		>
			<text>text</text>
		</condition>).toEqual({
			type: 'condition',
			title: [{type: 'text', value: 'Some title'}],
			status: 'success',
			children: [{type: 'text', value: 'text'}],
		});

		expect(<condition title={<fragment>Some <input>input</input> title</fragment>}>
			<text>text</text>
		</condition>).toEqual({
			type: 'condition',
			title: [
				{type: 'text', value: 'Some '},
				{type: 'input', value: 'input'},
				{type: 'text', value: ' title'},
			],
			children: [{type: 'text', value: 'text'}],
		});
	});

	it('should translate test line result line correctly', () => {
		expect(<test-line-result
			status="fail"
			message={<text>Some text</text>}
		>
			<test-line title={<text>Some title</text>}/>
		</test-line-result>).toMatchSnapshot();
		expect(<test-line-result
			status="fail"
			message={<text>Some text</text>}
			screenshot="some/path/to/screenshot.png"
			docs="//link/to/docs"
		>
			<test-line title={<text>Some title</text>}/>
		</test-line-result>).toMatchSnapshot();
	});

	it('should translate link correctly', () => {
		expect(<link href="http://some.url">link to some url</link>).toMatchSnapshot();
	});

	it('should throw error if test result get anything other then a single test line', () => {
		expect(() => <test-line-result status="success"><text>test</text></test-line-result>).toThrowError();
		expect(() => <test-line-result status="success">
			<test-line title={<text>test</text>}/>
			<test-line title={<text>test</text>}/>
		</test-line-result>).toThrowError();
	});

	describe('flatten util', () => {
		it('should flatten array of any depth', () => {
			expect(flatten([1, [2, [3, 4], 5], 6])).toEqual([1, 2, 3, 4, 5, 6]);
		});

		it('should always return an array, even if input is not', () => {
			expect(flatten('whatever')).toEqual(['whatever']);
		});
	});
});
