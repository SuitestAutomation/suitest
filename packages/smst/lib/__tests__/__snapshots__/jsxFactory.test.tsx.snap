// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsxFactory should create correct Prop element 1`] = `
Object {
  "actualValue": "actual value",
  "comparator": "=",
  "contentType": "inline",
  "expectedValue": Array [
    Object {
      "type": "text",
      "value": "expected value",
    },
  ],
  "name": Array [
    Object {
      "type": "text",
      "value": "prop name",
    },
  ],
  "status": "success",
  "type": "prop",
}
`;

exports[`jsxFactory should create correct Prop element 2`] = `
Object {
  "comparator": undefined,
  "contentType": "block",
  "expectedValue": Object {
    "language": "javascript",
    "type": "code-block",
    "value": "expected value",
  },
  "name": Array [
    Object {
      "type": "text",
      "value": "prop name",
    },
  ],
  "status": "fail",
  "type": "prop",
}
`;

exports[`jsxFactory should create correct Props element 1`] = `
Object {
  "children": Array [
    Object {
      "actualValue": undefined,
      "comparator": undefined,
      "contentType": "inline",
      "expectedValue": Array [
        Object {
          "type": "text",
          "value": "test",
        },
      ],
      "name": Array [
        Object {
          "type": "text",
          "value": "test",
        },
      ],
      "status": undefined,
      "type": "prop",
    },
  ],
  "type": "props",
}
`;

exports[`jsxFactory should translate test line result line correctly 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "title": Array [
        Object {
          "type": "text",
          "value": "Some title",
        },
      ],
      "type": "test-line",
    },
  ],
  "message": Array [
    Object {
      "type": "text",
      "value": "Some text",
    },
  ],
  "screenshot": undefined,
  "status": "fail",
  "type": "test-line-result",
}
`;

exports[`jsxFactory should translate test line result line correctly 2`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "title": Array [
        Object {
          "type": "text",
          "value": "Some title",
        },
      ],
      "type": "test-line",
    },
  ],
  "message": Array [
    Object {
      "type": "text",
      "value": "Some text",
    },
  ],
  "screenshot": "some/path/to/screenshot.png",
  "status": "fail",
  "type": "test-line-result",
}
`;
